utf8.c

```diff
@@ -3204,17 +3216,24 @@ see L</sv_recode_to_utf8>().
 */
 
 U8*
-Perl_bytes_to_utf8(pTHX_ const U8 *s, STRLEN *lenp)
+Perl_bytes_to_utf8_free_me(pTHX_ const U8 *s, Size_t *lenp,
+                                 const U8 ** free_me_ptr)
 {
-    const U8 * const send = s + (*lenp);
+    PERL_ARGS_ASSERT_BYTES_TO_UTF8_FREE_ME;
+    PERL_UNUSED_CONTEXT;
+
+    const U8 * const send = s + *lenp;
+    Size_t variant_count = variant_under_utf8_count(s, send);
+    if (free_me_ptr != NULL && variant_count == 0 && s[*lenp-1] == '\0') {
```

Comment:
Here it looks like `*lenp` includes the terminating NUL, but below it doesn't.

Consider when `*lenp` starts as 1, this expects `s[0]` to be NUL which doesn't match what I expect from this API.

That said, we would be assuming that `s[*lenp]` is valid, `s+*lenp` would always be valid as a one-past-the-end pointer, but such a pointer cannot be dereferenced.

So that `s[*lenp]` is safe becomes a pre-condition when `free_me_ptr` isn't NULL.

Reply:
I'm trying to understand this comment.  I don't see how I'm dereferencing `s[*lenp]`.  I am dereferencing `s[*lenp - 1]`.

The input is not required to be NUlL terminated, but the output is.  So if  `*lenp` is 1, it looks at `s[0]`.  If that is NUL, the function returns `s` unchanged, as it is a NUL-terminated string whose representation doesn't change when encoded in UTF-8.  If it isn't a NUL, the function allocates new memory that includes whatever byte is in `s[0]` and appends a NUL to it.

In re-reading the code, I see I failed to consider the possibility that `*lenp` is 0, and that I might be overallocating the new memory by 1 byte.  And I did a bit more clean up, so I dereference instead `*(send -1)`.  And I think it is better to dereference a pointer once into a local variable, rather than to dereference it multiple times

Reply:
I think you reviewed this before refreshing with the latest version available at the time.  I had already noticed the double NUL and fixed it.

I don't know what to do about the length disparity.  If you include the NUL in the length in blead, you will get a double NUL.  In order for the new form to know that there is a trailing NUL, it has to be able examine that byte, and so the length has to include it.  I added a paragraph to the pod explaining it. (hopefully)

Reply:
It has since occurred to me that it might be better to just not make the guarantee of a trailing NUL if there is no other reason to allocate new memory.